test/extended/oauth/oauthcertfallback.go

```diff
@@ -0,0 +1,151 @@
+)
+
+var _ = g.Describe("The OAuth server", func() {
+	defer g.GinkgoRecover()
+	oc := exutil.NewCLI("oauth", exutil.KubeConfigPath())
+
+	g.It("has the correct token and certificate fallback semantics", func() {
+		fakecadir, err := ioutil.TempDir("", "fakeca")
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		defer os.RemoveAll(fakecadir)
+
+		fakeca, err := (&admin.CreateSignerCertOptions{
+			CertFile:   path.Join(fakecadir, "fakeca.crt"),
+			KeyFile:    path.Join(fakecadir, "fakeca.key"),
+			SerialFile: path.Join(fakecadir, "fakeca.serial"),
+			Name:       "foo",
+			IOStreams:  genericclioptions.NewTestIOStreamsDiscard(),
+			Overwrite:  true,
+		}).CreateSignerCert()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		clientCertConfig, err := fakeca.MakeClientCertificate(
+			path.Join(fakecadir, "fakeclient.crt"),
+			path.Join(fakecadir, "fakeclient.key"),
+			&user.DefaultInfo{Name: "fakeuser"},
+			365*2, /* 2 years */
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		invalidCert.CertData, invalidCert.KeyData, err = clientCertConfig.GetPEMBytes()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		tokenUser := fmt.Sprintf("system:serviceaccount:%s:builder", oc.Namespace())
+
+		validToken, err := oc.Run("sa").Args("get-token", "builder").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(validToken).ToNot(o.BeEmpty())
+
+		adminConfig := oc.AdminConfig()
+		validCert := adminConfig.TLSClientConfig
+
+		for name, test := range map[string]struct {
+			token         string
+			cert          rest.TLSClientConfig
+			expectedUser  string
+			errorExpected bool
+			errorString   string
+		}{
+			"valid token, valid cert": {
+				token:        validToken,
+				cert:         validCert,
+				expectedUser: tokenUser,
+			},
+			"valid token, invalid cert": {
+				token:        validToken,
+				cert:         invalidCert,
+				expectedUser: tokenUser,
+			},
+			"valid token, no cert": {
+				token:        validToken,
+				cert:         noCert,
+				expectedUser: tokenUser,
+			},
+			"invalid token, valid cert": {
+				token:         invalidToken,
+				cert:          validCert,
+				errorExpected: true,
+				errorString:   unauthorizedError,
+			},
+			"invalid token, invalid cert": {
+				token:         invalidToken,
+				cert:          invalidCert,
+				errorExpected: true,
+				errorString:   unauthorizedError,
+			},
+			"invalid token, no cert": {
+				token:         invalidToken,
+				cert:          noCert,
+				errorExpected: true,
+				errorString:   unauthorizedError,
+			},
+			"no token, valid cert": {
+				token:        noToken,
+				cert:         validCert,
+				expectedUser: certUser,
+			},
+			"no token, invalid cert": {
+				token:         noToken,
+				cert:          invalidCert,
+				errorExpected: true,
+				errorString:   unauthorizedError,
+			},
+			"no token, no cert": {
+				token:         noToken,
+				cert:          noCert,
+				errorExpected: true,
+				errorString:   anonymousError,
+			},
+		} {
```

Comment:
Instead of passing `name` to every `Expect` it is probably better to have each iteration be a `g.Describe(name, ...`